<a href="https://colab.research.google.com/github/bayan2002/Prediction-of-Product-Sales/blob/main/Data_Cleaning_Core_Starter_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning Core [Starter Notebook]



- Student Name: Bayan Abd El Bary

## The Data

You will be working with a modified version of the Titanitc dataset:

- [(Modified) Titanic Data](https://drive.google.com/file/d/1Uw7RzQMYSARCFnzJQqqNlZpUouXr16-8/view?usp=sharing)

- [Data Dictionary](https://docs.google.com/document/d/1uEeANvKURYmu-HUDGzbGlMmvAUwooatUg-gHUCDMh-Y/edit?usp=sharing)

___

- Save the files to your Course 1 > Week 2 > Data Folder.


## A) Load and Preview Data


- Mount google drive and navigate to the file on the Files sidebar.


In [1]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
from google.colab import drive
drive.mount('/content/drive')


- Navigate to the file on the Files side bar and copy the path
- Paste it below and save it as a variable called "fpath"

In [3]:
# Import required packages
import pandas as pd
# Save the filepath and load the data as "df_titanic"
fpath= "/content/drive/MyDrive/AXSOSACADEMY/01-Fundamentals/Week02/Data/titanic_dirty.csv"
df_titanic= pd.read_csv(fpath)
df_titanic

,Unnamed: 0,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,$7.25,NaN,S
1,1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,$71.28,C85,C
2,2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,$7.93,NaN,S
3,3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,$53.10,C123,S
4,4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,$8.05,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...,...
901,622,623,1,3,"Nakid, Mr. Sahid",male,20.0,1,1,2653,$15.74,NaN,C
902,77,78,0,3,"Moutal, Mr. Rahamin Haim",male,NaN,0,0,374746,$8.05,NaN,S
903,715,716,0,3,"Soholt, Mr. Peter Andreas Lauritz Andersen",male,19.0,0,0,348124,$7.65,F G73,S
904,623,624,0,3,"Hansen, Mr. Henry Damsgaard",male,21.0,0,0,350029,$7.85,NaN,S


In [5]:
# Display the first 5 rows
df_titanic.head()

,Unnamed: 0,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,$7.25,NaN,S
1,1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,$71.28,C85,C
2,2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,$7.93,NaN,S
3,3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,$53.10,C123,S
4,4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,$8.05,NaN,S


In [4]:
# Display the last 5 rows
df_titanic.tail()

,Unnamed: 0,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
901,622,623,1,3,"Nakid, Mr. Sahid",male,20.0,1,1,2653,$15.74,NaN,C
902,77,78,0,3,"Moutal, Mr. Rahamin Haim",male,NaN,0,0,374746,$8.05,NaN,S
903,715,716,0,3,"Soholt, Mr. Peter Andreas Lauritz Andersen",male,19.0,0,0,348124,$7.65,F G73,S
904,623,624,0,3,"Hansen, Mr. Henry Damsgaard",male,21.0,0,0,350029,$7.85,NaN,S
905,888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,$23.45,NaN,S


- Display summary information (including dtypes, number of rows, number of columns).


In [7]:
# Display the info for the dataframe
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 906 entries, 0 to 905
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   906 non-null    int64  
 1   PassengerId  906 non-null    int64  
 2   Survived     906 non-null    int64  
 3   Pclass       906 non-null    int64  
 4   Name         906 non-null    object 
 5   Sex          906 non-null    object 
 6   Age          723 non-null    float64
 7   SibSp        906 non-null    object 
 8   Parch        906 non-null    int64  
 9   Ticket       906 non-null    object 
 10  Fare         906 non-null    object 
 11  Cabin        207 non-null    object 
 12  Embarked     904 non-null    object 
dtypes: float64(1), int64(5), object(7)
memory usage: 92.1+ KB


**Scan the .info() to answer the following questions about the data:**

>Add your answers in this Markdown cell, right below each question:

**Q: A1) How many rows are there?**
- Your Answer:
  - 906

**Q: A2) How many columns are there?**
- Your Answer:
  - 13




**Q: A3)How many numeric columns?**
- Your Answer:
  - 6

**Q: A4) How many categorical/string columns are there (dtype='object')?**
- Your Answer:
  - `7`

## B) Check Data Types and Convert (if required)


> Notice there are several columns with "object" datatype.

- **Use select_dtypes to view only the object columns**


In [8]:
# Use .select_dtypes to view only object columns
df_titanic.select_dtypes('object')

,Name,Sex,SibSp,Ticket,Fare,Cabin,Embarked
0,"Braund, Mr. Owen Harris",male,1,A/5 21171,$7.25,NaN,S
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,1,PC 17599,$71.28,C85,C
2,"Heikkinen, Miss. Laina",female,0,STON/O2. 3101282,$7.93,NaN,S
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,1,113803,$53.10,C123,S
4,"Allen, Mr. William Henry",male,0,373450,$8.05,NaN,S
...,...,...,...,...,...,...,...
901,"Nakid, Mr. Sahid",male,1,2653,$15.74,NaN,C
902,"Moutal, Mr. Rahamin Haim",male,0,374746,$8.05,NaN,S
903,"Soholt, Mr. Peter Andreas Lauritz Andersen",male,0,348124,$7.65,F G73,S
904,"Hansen, Mr. Henry Damsgaard",male,0,350029,$7.85,NaN,S


- Inspect the values in each column to confirm if they should be converted to numeric dtypes and answer the following questions:




**Q: B1) Which columns should be converted to a numeric dtype?**  (Hint: there are 2 columns that need to be converted)
- **Your Answer:**
  - `SibSp`
  - `Fare`
    




### Convert Dtypes

#### Convert Fare

In [12]:
# Inspect the Fare column's value counts
df_titanic['Fare'].value_counts(dropna=False)


,count
Fare,
$8.05,45
$13.00,42
$7.90,38
$7.75,35
$26.00,31
...,...
$14.11,1
$50.50,1
$5.00,1




>In a Markdown cell below your displayed info, answer the following questions:

**Q: B2)  Do you see any values that are string characters that need to be replaced?**
- Your Answer:
  - `$`

**Q: B3) If so, do you need to replace a single part of a string (`df[col].str.replace`) or an entire value (`df[col].replace`)?**
- Your Answer:
  - `df[col].str.replace`

**Perform the necessary steps to convert to numeric dtype:**

- 1) Replace/remove the non-numeric values.

In [15]:
# Remove  the "$" from fare column by replacing it with nothing
df_titanic['Fare'].str.replace('$',"")

# Check the .value_counts to confirm the change
df_titanic['Fare'].value_counts()

,count
Fare,
$8.05,45
$13.00,42
$7.90,38
$7.75,35
$26.00,31
...,...
$14.11,1
$50.50,1
$5.00,1



- 2) Convert the column to a float dtype.

In [ ]:
# Change the datatype of Fare column to float

# Confirm the datatype


#### Convert SibSp

In [ ]:
# Inspect the Sibsp column's value counts


>In a Markdown cell below your displayed info, answer the following questions:

**Q: B3)  Do you see any values that are string characters that need to be replaced?**
- Your Answer:
  - `...`

**Q: B4) If so, do you need to replace a single part of a string (`df[col].str.replace`) or an entire value (`df[col].replace`)?**
- Your Answer:
  - `...`

**Perform the necessary steps to convert to numeric dtype:**

- 1) Replace/remove the non-numeric values.

In [ ]:
# There are 3 values of "one" which should be replaced with "1"

# Display the value counts again to confirm the change



- 2) Convert the column to a float dtype.

In [ ]:
# Convert the SibSp column to integer dtype

# Confirm the datatype


## C) Drop Unwanted Columns

- Drop the 'Unnamed: 0' column because it is an error

In [ ]:
# Drop the Unnamed: 0 column


# Display the .head to confirm the change


## D) Rename Columns
(that have Confusing/Abbreviated Column Names)

In the cell below:
- Rename the SibSp column as "Siblings or Spouses"
- Rename the Parch as "Parents or Children"

In [ ]:
# Create a dictionary called rename_dict
# with current column names as keys
# and new replacement names as the values.


In [ ]:
# Use the rename_dict to change the column names in the dataframe

# Check the .head() to confirm the nams changed.


## E) Duplicate Rows

### Identifying Duplicates


- Use `df.duplicated` to identify if there are any duplicate rows in the data.


In [ ]:
# Save a filter called "duplicated_rows" that identifies duplicated rows


# Get the sum of duplicated_rows filter to see the number of duplicate rows



**Q: E1) How many duplicate rows are in the data?**
- Your Answer:
  - `...`

  
- Visualize all copies of the duplicate rows using df.duplicated(keep=False).
  - Sort the rows by the PassengerID column to view duplicate rows together.

In [ ]:
# Save a new filter called "duplicated_rows_all" that identifies ALL
# of the duplicated rows (including the first). Hint: use the "keep" argument.


# Calculate the sum of the duplicated_rows_all filter


In [ ]:
# View all duplicated rows using your duplicated_rows_all filter
# Sort by PassengerId


In [ ]:
# Drop duplicates (but keep the first!)

# Obtain the sum of duplicated to confirm all have been dropped


## F) Unique Identifiers
- Find and Drop Columns with Unique Values for Every Row.

In [ ]:
# Use .nunqiue() to get # of unique values for each column and save as a variable.


In [ ]:
# Use the result to determine if any of the columns have 100% unique values.


**Q: F1) Which 2 features have 100% unique values in every row?**
- Your Answer:
  - Column 1: `...`
  - Column 2: `...`




>Generally, a column that is NOT the index, and has a unique value for each row will not be helpful for analysis and should be dropped from the data.

- Inspect the two columns with 100% unique values:


In [ ]:
# Slice the two columns with 100% unique values to view



- **Of the two columns with 100% unique values:**
  - Set the numeric column as the DataFrame's index.
  - Drop the string column from the DataFrame.


In [ ]:
# Set the numeric column as the DataFrame's index.


In [ ]:
# Drop the string column from the DataFrame.


___

## G) Missing Values

In [ ]:
# Get the number of null values for all columns and save as a variable


In [ ]:
# Use this result to calculate the percentage of null values for each column.


In [ ]:
# Visualize the missing values using the missingno package's matrix plot.


#### Address the null values:

- Fill the null values in categorical columns with a placeholder category called MISSING.


In [ ]:
# save list of categorical column name. (Hint: select_dtypes)


In [ ]:
# use the list of cat columns to fill null values with MISSING



- Fill the null values in numeric columns with a placeholder value of -1



In [ ]:
# save list of numeric column names (Hint: select_dtypes)


In [ ]:
# use the list of num columns to fill null values with -1


- Use either missingno's matrix plot or Pandas to confirm that there are 0 null values remaining.


In [ ]:
# Confirm no more nulls using Pandas


In [ ]:
# Confirm no more nulls using using missingno


## H) Consistent Values

- Check each string column's value counts for inconsistent values.
  - E.g. "Cat" vs "cat", "Y" vs. "Yes"



- Fix/replace the inconsistent values with the values you believe are correct.


- Display the value_counts again to confirm the values are now consistent.
  

In [ ]:
# Save a list of object columns (Hint: select_dtypes)


In [ ]:
# Check the nunique for just the object cols


- Note that out of the 4 categorical columns remaining:
  - 2 columns have a small number of categories  (Sex and Embarked)
  - 2 columns have over 100 categories (Ticket and Cabin).

- We do not want to keep categorical features with so many categories.


- **Drop the Ticket column and Cabin column.**

In [ ]:
# Drop the columns listed above


### For Each Remaining Categorical Column:

- Display the value counts for the column
- Take a note of any values that are not consistent.
- Remove/fix the inconsistent values
- Confirm by checking the value counts again

In [ ]:
# Remake the list of string columns (after dropping previous cols)

# loop through the list of string columns

  # print the value counts for the column

  # Print an empty line for readability


**Q: H1) Which columns have inconsistent values that need to be addressed?**
- Your Answer:
  - `...`
  - `...`

### Checking & Fixing Consistency: Column 1

In [ ]:
# Check for inconsistencies by inspecting the value_counts for Embarked


In [ ]:
# Replace "Queenstown" with "Q"

# Check the value counts again to confirm


### Checking and Fixing Consistency: Column 2

In [ ]:
# # Check for inconsistencies by inspecting the values_counts for Sex


In [ ]:
# Replace "F" with "female"


# Check the value counts again to confirm


## I) Save the Final DataFrame to Csv

- Save the cleaned DataFrame to your Google Drive in the CodingDojo>  01-Fundamentals> Week02>Data folder.
  - Name the file "data-cleaning-core.csv"


In [ ]:
# Save the folder + filename as a variable called fpath


# Use to_csv to save the dataframe using the filename


- Confirm the data was saved correctly.

In [ ]:
# Read in your new .csv file with Pandas.

# Display the info summary.

# Display the first 5 rows.


### Save and Submit your work

When the assignment is complete, download your notebook as a .ipynb file and submit it on the learning platform.  